### DISCUSSION TITLE
Why do docker images need to support amd64 and arm64?

### DISCUSSION THREAD

**Discussion by rzuckerm** [PlummersSoftwareLLC/Primes#951] [Q&A] [answered]
I'm a bit confused about the [requirement](https://github.com/PlummersSoftwareLLC/Primes/blob/drag-race/CONTRIBUTING.md#support-for-hardware-architectures) that docker images support both amd64 and arm64. I understand the part about indicating that only a specific architecture is supported. The reason why I ask is that I noticed that in the CI workflow [here](https://github.com/PlummersSoftwareLLC/Primes/blob/drag-race/.github/workflows/CI.yml#L56-L68) and [here](https://github.com/PlummersSoftwareLLC/Primes/blob/drag-race/.github/workflows/CI.yml#L91-L111), the arm64 image only gets built if the `arch-arm64` file exists. Otherwise, the amd64 image is built. It seems like both images should get built when both files do not exist or both files do exist. In other words, it seems like it should work like this:

| `arch-amd64` exists | `arch-arm64` exists | build amd64 | build arm64 |
| --------------------- | --------------------- | -------------- | ------------- |
| N                             | N                            | Y                    | Y                  |
| N                             | Y                            | N                    | Y                  |
| Y                             | N                            | Y                    | N                  |
| Y                             | Y                            | Y                     | Y                  |

Please don't interpret this as arguing with the requirement. I'm just trying to understand it.

**Comment by rbergen** [comment, marked as answer]
The two things are basically mostly independent. 

CI is there to confirm the solutions' Docker images build, and building for AMD64 is a lot faster than ARM64 - as that architecture is emulated using QEMU during the CI workflow. So, if an AMD64 base image is present then we prefer that for CI, but still "test" the Dockerfile if ARM64 is the only option. We acknowledge that this doesn't cover the scenario where an AMD64 Docker build behaves differently from an ARM64 Docker build _based on the same base image_, partly because we've found that is very rare indeed.

The "requirement" for the ability for solutions to run on ARM64 if possible is rooted in the facts that

1. An increasing amount of workloads on the planet are being run on ARM64
2. The performance of a programming language/approach on that architecture is therefore relevant, and
3. Any difference of performance between the AMD64 and ARM64 architectures is therefore interesting also.

Now, the actual "daily" comparison between the two architectures' performance is complicated by the fact that we currently have no ARM64 daily benchmark runners. That's a practical limitation that's unfortunate, but I don't think negates the arguments I just mentioned for the desire to support ARM64 as such.
